updates to comet integration

DN6 opened the pull request:
This PR:

1. Adds `Comet from` source tracking to YOLOv8
2. Adds support for logging training and validation images and model checkpoints every `N` epochs during training
3. Adds support for logging interactive Confusion Matrix to Comet
4. Adds support for Online and Offline Experiment logging

TODO:

1. Test logging from Distributed Training Runs
2. Investigate strange issue where certain assets are logged twice with the same step when training ends (Confusion Matrix, Plots, and Checkpoints)   

Example Experiment:
https://www.comet.com/team-comet-ml/test-yolov8-demo/4936d49f31d54aa086509a8156a8c91d

DN6 committed:
updates to comet integration

haarcuba commented on ultralytics/yolo/utils/callbacks/comet.py:
```diff
@@ -1,7 +1,12 @@
 from ultralytics.yolo.utils.torch_utils import get_flops, get_num_params
 
+COMET_MODE = os.getenv('COMET_MODE', 'online')
+MODEL_NAME = os.getenv('COMET_MODEL_NAME', 'YOLOv8')
```
the `MODEL_NAME` constant appears to be unused. Is it needed?

haarcuba commented on ultralytics/yolo/utils/callbacks/comet.py:
```diff
@@ -11,43 +16,140 @@
+
+def _log_confusion_matrix(trainer, experiment, curr_epoch):
+    conf_mat = trainer.validator.confusion_matrix.matrix
+    # Ensure dictionary with label names is sorted based on keys
```
`.items()` yields `(key, value)` tuples, so sorting them will already produce sorting by key. So, no need for the `key=lambda x: x[0]`, just `sorted` will do.

haarcuba commented on ultralytics/yolo/utils/callbacks/comet.py:
```diff
@@ -11,43 +16,140 @@
-        LOGGER.warning(f'WARNING ⚠️ Comet installed but not initialized correctly, not logging this run. {e}')
+    # Ensures that the experiment object is created in a single process
+    # during distributed training.
+    if RANK in {-1, 0}:
```
why `{-1, 0}` and not `(-1, 0)`? it's harder to understand the order of iteration this way.

haarcuba commented on ultralytics/yolo/utils/callbacks/comet.py:
```diff
@@ -11,43 +16,140 @@
+            experiment.log_parameters(vars(trainer.args))
+            experiment.log_other('Created from', 'yolov8')
+        except Exception as e:
+            LOGGER.warning(f'WARNING ⚠️ Comet installed but not initialized correctly, not logging this run. {e}')
```
who know what the true reason is - I don't think that you can know that Comet is "not initialized correctly" - IMO replace this with a more generic error message

haarcuba commented on ultralytics/yolo/utils/callbacks/comet.py:
```diff
@@ -11,43 +16,140 @@
+            epoch=curr_epoch,
+        )
+
+        for f in trainer.save_dir.glob('train_batch*.jpg'):
```
`f` is not a proper name, use `file` or `filename` or `path` or whatever it actually is.

haarcuba commented on ultralytics/yolo/utils/callbacks/comet.py:
```diff
@@ -11,43 +16,140 @@
+    curr_step = metadata['curr_step']
+    save_assets = metadata['save_assets']
+
     if experiment:
```
it will be easier to read this if you do the opposite:

```
if not experiment:
   return

experiment.log_metrics(...)
# rest of code
```

see [Avoid Nesting Important Flows](https://github.com/PracticeFoxyCode/practice#avoid-nesting-important-flows---prefer-to-nest-unimportant-flows)

haarcuba commented on ultralytics/yolo/utils/callbacks/comet.py:
```diff
@@ -11,43 +16,140 @@
+                'model/speed(ms)': round(trainer.validator.speed['inference'], 3),}
+            experiment.log_metrics(model_info, epoch=curr_epoch)
+
+        if save_assets:
```
this function is too long, extract this part

```python
if save_assets:
       _save(experiment, trainer)

# in another place
def _save(experiment, trainer):
     # actual code that saves it

```

haarcuba commented on ultralytics/yolo/utils/callbacks/comet.py:
```diff
@@ -11,43 +16,140 @@
+    metadata = _fetch_run_metadata(trainer)
+    curr_epoch = metadata['curr_epoch']
+
+    if experiment:
```
replace with

```python
if not experiment:
    return

...
```
see: [Avoid Nesting Important Flows](https://github.com/PracticeFoxyCode/practice#avoid-nesting-important-flows---prefer-to-nest-unimportant-flows)

haarcuba commented on ultralytics/yolo/utils/callbacks/comet.py:
```diff
@@ -11,43 +16,140 @@
+        _log_confusion_matrix(trainer, experiment, curr_epoch)
+
+        try:
+            for plots in [
```
extract this to another function that has to do with plots.

haarcuba commented on ultralytics/yolo/utils/callbacks/comet.py:
```diff
@@ -11,43 +16,140 @@
 
 
-callbacks = {
+callbacks = ({
```
this `if-else` is hard to read. Use the longer form.

```
_actual_callbacks = {
   # the big dict here
}

if comet_ml:
   callbacks = {}
else:
  callbacks = _actual_callbacks

DN6 committed:
add support for logging image predictions, and confusion matrix

Lothiraldan commented on ultralytics/yolo/utils/callbacks/comet.py:
```diff
@@ -11,43 +37,308 @@
+    indices = batch['batch_idx'] == img_idx
+    bboxes = batch['bboxes'][indices]
+    if len(bboxes) == 0:
+        LOGGER.warning(f'COMET WARNING: Image: {image_path} has no bounding boxes labels')
```
Can the user do something to "fix" that warning? Could that warning be printed for each image that we are trying to logging?

DN6 replied:
It's unlikely that this is the case, but it is possible.  To suppress the warning, the user would have to change the `logging` level. @Lothiraldan think this should be changed to a `LOGGER.debug` message instead?

DN6 replied:
Or the user can remove the problematic images from the dataset.

Lothiraldan commented on ultralytics/yolo/utils/callbacks/comet.py:
```diff
@@ -11,43 +37,308 @@
+
+    predictions = metadata.get(image_id)
+    if not predictions:
+        LOGGER.warning(f'COMET WARNING: Image: {image_path} has no bounding boxes predictions')
```
Can the user do something to "fix" that warning? Could that warning be printed for each image that we are trying to logging?

DN6 committed:
add max image predictions

DN6 committed:
update comet utms

DN6 committed:
reset image prediction count when training ends

glenn-jocher committed:
`ultralytics 8.0.62`

dh031200 committed:
Fix callbacks by reference (#1783)

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Glenn Jocher <glenn.jocher@ultralytics.com>

MagicCodess committed:
fix onnx example code to get same results (#1780)

Co-authored-by: Glenn Jocher <glenn.jocher@ultralytics.com>

dh031200 committed:
Add option for plot (#1732)

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Glenn Jocher <glenn.jocher@ultralytics.com>
Co-authored-by: ayush chaurasia <ayush.chaurarsia@gmail.com>

DN6 committed:
Merge branch 'main' into comet-integration-updates

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit suggestions (#1798)

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Glenn Jocher <glenn.jocher@ultralytics.com>

AyushExel committed:
Generalize is_cli for track (#1802)

Co-authored-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
HUB train from `YOLO(APIKEY_MODELID)` support (#1793)

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

glenn-jocher committed:
Update comet.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

glenn-jocher committed:
Update Comet UTM link in both places in English README

Linked https://bit.ly/yolov8-readme-comet to https://www.comet.com/site/lp/yolov5-with-comet/?utm_source=yolov8&utm_medium=partner&utm_content=github

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

glenn-jocher committed:
Update 2 Comet links in README.zh-CN.md

@DN6

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Amjad50 committed:
Fix issue where settings yaml can be `None` in yolo/utils (#1797)

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
Update ci.yaml

glenn-jocher committed:
Merge branch 'updates' into comet-integration-updates

glenn-jocher committed:
Cleanup
